### PR TITLE
Feature 7 add restaurant page

### PR DESCRIPTION
**Pull request by vitaliimak**
Don't pay attention to styles, html, etc. They don't work correctly.
Could you look did I add axios library well?

### CODE REVIEW

**Review by r-yevchuk — DISMISSED** [review]

File: client/src/components/restaurant/restaurant.js
Comment (VanderbiGit) [inline client/src/components/restaurant/restaurant.js]: We use tab?

File: client/src/services/api.js
Comment (VanderbiGit) [inline client/src/services/api.js:10]: axios work

**Review by VanderbiGit — COMMENTED** [review]
you need connect to server

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: Yes. According to design_decisions and mock-ups channels

**Comment by vitaliimak** [comment]
> you need connect to server

I think, it's because of

> quote from slack
> Я спробував додати **фейк** бекенд

File: server/src/main/java/com/lunchtime/service/RestaurantService.java
Comment (vitaliimak) [inline server/src/main/java/com/lunchtime/service/RestaurantService.java:45]: It doesn't look very good. Can someone know a better solution?

**Comment by vitaliimak** [comment]
I think server part is ready to review.

**Comment by vitaliimak** [comment]
Frontend part is ready to review too

File: client/src/components/restaurant/restaurant.js
Comment (narekacy) [inline client/src/components/restaurant/restaurant.js]: "title" should be "**Feedback**", a little definition mistake

File: client/src/components/restaurant/restaurant.js
Comment (narekacy) [inline client/src/components/restaurant/restaurant.js]: I think it is better to try to `setState()` before checking if response has got any error. I would switch lines 25 and 30 placing. 

Of course, we should check first `if(!response.error)` (or if we get exact response that is needed for `setState` method) and only after this `else`.

File: client/src/components/restaurant/restaurant.js
Comment (narekacy) [reply client/src/components/restaurant/restaurant.js]: I guess we need something like **Tab**-"wrapper" (or even **RestaurantTab**) component with props due to React.js. It will also help our code to be easier.

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: I set `isFetching = false` in constructor so nested component won't break. If request returns error than `response.data = undefined` and we don't need to update state before if statement

File: server/src/main/java/com/lunchtime/controllers/RestaurantResource.java
Comment (narekacy) [inline server/src/main/java/com/lunchtime/controllers/RestaurantResource.java]: Since Spring 4.3 it is recommended to avoid using `@Autowired` annotation but use **final** field (`private final RestaurantService restaurantService` in our case) and injecting by constructor.

```
public RestaurantResource (RestaurantService restaurantService){
        this.restaurantService= restaurantService;
    }
```

File: client/src/components/restaurant/restaurant.js
Comment (narekacy) [reply client/src/components/restaurant/restaurant.js]: I like your way. It seems good.

On the same time I guess we will have more successful

Comment (narekacy) [reply client/src/components/restaurant/restaurant.js]: I like your way. It looks good.

At the same time I guess we'll have more successful responses rather than failed. That's why I would prefer to check ones but not twice. 

**EDITED**:
to check what we will use for our purpose

File: server/src/main/java/com/lunchtime/controllers/RestaurantResource.java
Comment (narekacy) [inline server/src/main/java/com/lunchtime/controllers/RestaurantResource.java:44]: I think you meant `if (restaurant.getId() == null)` in line 33

File: server/src/main/java/com/lunchtime/service/RestaurantService.java
Comment (narekacy) [reply server/src/main/java/com/lunchtime/service/RestaurantService.java:45]: For sure, Hibernate has prepared something useful for us. [**Check this**](https://stackoverflow.com/questions/2398981/effective-entity-update-in-hibernate), please

**Review by Andrii35 — COMMENTED** [review]
By creating of entities we need to consider the relationships between them.  You have in Restaurant entity two foreign keys "menu_id" and "owner_id".  I found information, that we need to use annotations "@OneToOne", "@OneToMany" or "@ManyToMany". How to do it you can read here https://www.javaworld.com/article/2077819/understanding-jpa-part-2-relationships-the-jpa-way.html or  https://easyjava.ru/data/jpa/jpa-i-svyazi-mezhdu-obektami/

**Comment by vitaliimak** [comment]
> By creating of entities we need to consider the relationships between them...

You are right, but in this task I do only restaurant part, I can't add these relations because tables "user", "menu", etc don't exist. Person who creates these entities should do it.

So I added temporary fake fields

File: server/src/main/java/com/lunchtime/service/RestaurantService.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/service/RestaurantService.java:45]: Thanks for answer. But `dynamicUpdate=true` doesn't help in this case. 
This code looks bad because developer must look after 2 places in codebase. When entity Restaurant is updated then method update(restaurant) must be updated too. 
I can't save newRestaraunt object directly, because it comes from api request and can contain fields like createdAt with wrong data (or other wrong fields).

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: Sorry, I'm not sure that understand you

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: In this case
```
then((response) => {
        this.setState({
          restaurant: response.data,
          isFetching: true,
        });
        if (response.error) {
          // eslint-disable-next-line no-console
          console.error(response);
          this.setState({
            restaurant: {},
            isFetching: false,
          });
          // return;
        }
      });
```
I have this result if error happens
image

I use Error-First principle to handle errors http://fredkschott.com/post/2014/03/understanding-error-first-callbacks-in-node-js/
It doesn't affected user experience because if statement works very fast

But anyway, could you show code example with your solution?

File: server/src/main/java/com/lunchtime/controllers/RestaurantResource.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/controllers/RestaurantResource.java]: Thanks, fixed

File: server/src/main/java/com/lunchtime/controllers/RestaurantResource.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/controllers/RestaurantResource.java:44]: No if new Restaurant has id (restaurant.getId() != null) then user must use update request, because client can't know id in new Restaurant entity. Only database creates id in every new entities

File: server/src/main/java/com/lunchtime/service/RestaurantService.java
Comment (narekacy) [reply server/src/main/java/com/lunchtime/service/RestaurantService.java:45]: I think we should hold this place for future and return here. There must be more flexible way to update our entity. Now we can't imagine Hibernate's power, but we will consider to improve it later. It's ok.

File: client/src/components/restaurant/restaurant.js
Comment (narekacy) [reply client/src/components/restaurant/restaurant.js]:
```
      .then((response) => {
        if (!response.error) {
            this.setState({
                restaurant: response.data,
                isFetching: true,
            });
            return;
        } else {
            console.error(response);
        }
      });
```

File: server/src/main/java/com/lunchtime/controllers/RestaurantResource.java
Comment (narekacy) [reply server/src/main/java/com/lunchtime/controllers/RestaurantResource.java:44]: Got it. Thanks!

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: I try to avoid negative conditionals https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29#understandability-tips

It makes code less readable

**Comment by VanderbiGit** [comment]
Screenshot from 2020-03-10 11-41-55

I have some error. In your computer work well?

**Comment by vitaliimak** [comment]
> I have some error. In your computer work well?

It's because of very magic annotations `@Setter and @Getter` in Restaurant class. Those annotations generate getters and setters in runtime and IDEA doesn't know about those

Code compiles well. But you can fix those errors in ide. Just install plugin https://plugins.jetbrains.com/plugin/6317-lombok and update settings https://stackoverflow.com/questions/17729384/lombok-added-but-getters-and-setters-not-recognized-in-intellij-idea

**Comment by VanderbiGit** [comment]
What about frontend?
Screenshot from 2020-03-10 11-53-10

**Comment by vitaliimak** [comment]
> What about frontend?

I hardcoded restaurantId  from my databse. This id should send dynamically from restaurant list. It will do other developer. As you can see, even when id doesn't exist application doesn't crash and does show `Loading...` message.

You can create new Restaurant entity in your database (from Postman) and update restaurantId in restaurant-list.js for testing

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (narekacy) [inline server/src/main/java/com/lunchtime/models/Restaurant.java:41]: There is `@URL` annotation for field's validation as a website. But we should take care about its regex: 

_Per default the constraint validator for this constraint uses the java.net.URL constructor to validate the string. This means that a matching protocol handler needs to be available. Handlers for the following protocols are guaranteed to exist within a default JVM - http, https, ftp, file, and jar. See also the Javadoc for URL._  -> [link](https://softwareengineering.stackexchange.com/questions/356540/how-to-represent-and-validate-website-urls-for-a-jpa-entity)

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (narekacy) [inline server/src/main/java/com/lunchtime/models/Restaurant.java]: Due to JavaBean Specifications (**8.3.2 Boolean properties**), we should use:
public boolean (or Boolean) **is**Deleted<PropertyName>() instead of **getIs**Deleted<PropertyName>()

Also we can use a ternary operator for code's readability

**Review by VanderbiGit — DISMISSED** [review]

**Review by Andrii35 — DISMISSED** [review]

**Review by r-yevchuk — DISMISSED** [review]

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/models/Restaurant.java:41]: Thanks, very useful annotation! But by default it forbids urls like `somelink.domain.com` (without protocol). It can be customizable by regex, as I understand.
As our business process requires be moderator who check restaurant entity before approve. I think this annotation will bring more complexity then advantage. E.g. if this field must be valid there is must to be validation in input in frontend
I also didn't add annotation `@NotBlank` as restaurant may not have website

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/models/Restaurant.java]: Unfortunately I didn't find this rule, but I think you're right

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (narekacy) [reply server/src/main/java/com/lunchtime/models/Restaurant.java:41]: I think we should embody possible complexity to advantage :) Of course we should add regex so that user can see that his web-URL is not valid, but not moderator.

i.e.: coffee-cafe.com (where first '**c'** is not latin one: this can happen when user starts to type and doesn't switch keyboard language. So, when he notices this he backspace his not-latin letters up to 'c', thinking that 'c' on cyrillic and 'c' on latin is the same symbol :D

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (narekacy) [reply server/src/main/java/com/lunchtime/models/Restaurant.java]: Thank you for reliance! :)

**[Here ](https://download.oracle.com/otndocs/jcp/7224-javabeans-1.01-fr-spec-oth-JSpec/)** you can download it and see **8.3.2** paragraph. I also attached an image from this file, supporting this recommendation.

t-bp

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/models/Restaurant.java:41]: Regular expressions are very powerful feature and very dangerous and complicated. I think we need to review our requirements according to this conversation and will decide what way is best

There are few references about email:
https://habr.com/en/post/175375/
https://habr.com/en/post/274985/

There is RFC about website naming (I didn't read):
https://tools.ietf.org/html/rfc819

Comment (narekacy) [reply server/src/main/java/com/lunchtime/models/Restaurant.java:41]: I could be wrong. I am going to maintain your opinion. It will be the best decision to review our requirements.

When reviewing, let us remember that this "micro" feature can add point to our developer skills but only if we use smart solution in short time:)

Anyway, may be this "micro" feature is not worth this conversation?)

Sorry in advance :)

File: server/src/main/java/com/lunchtime/models/Restaurant.java
Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/models/Restaurant.java]: I am completely confused)
Should I rename property `private Boolean isDeleted` to `private Boolean deleted`?
Because pattern `get<PropertyName>` where `PropertyName` is `isDeleted` was realized in previous code iteration

Comment (narekacy) [reply server/src/main/java/com/lunchtime/models/Restaurant.java]: If I understand you correctly (line 19), I would keep property `private Boolean isDeleted` and rename`getIsDeleted()` to `isDeleted()`. And we will have

```
private Boolean isDeleted
public Boolean isDeleted()
```

Comment (vitaliimak) [reply server/src/main/java/com/lunchtime/models/Restaurant.java]: Fixed

File: client/src/components/restaurant/restaurant.js
Comment (vitaliimak) [reply client/src/components/restaurant/restaurant.js]: @narekacy Could you explain it better?

Comment (narekacy) [reply client/src/components/restaurant/restaurant.js]: Check this [video tutorial](https://www.youtube.com/watch?v=-mDs48HB3II&list=PLcvhF2Wqh7DNVy1OCUpG3i5lyxyBWhGZ8&index=16), please. You may set 1.5 or even 2.0 playback speed, this guy explains it in excellent way. 

When you finish watching it, remind yourself that you also can use an array to pass in it your component as props with `map()` magic.

**Review by VanderbiGit — APPROVED** [review]

**Review by Andrii35 — APPROVED** [review]

**Review by r-yevchuk — APPROVED** [review]

**Pull request merged**
